src/pubsub/ua_pubsub_reader.c

```diff
@@ -248,6 +249,29 @@ UA_DataSetReader_generateNetworkMessage(UA_PubSubConnection *pubSubConnection, U
                                  (u64)UA_UADPNETWORKMESSAGECONTENTMASK_DATASETCLASSID) != 0;
     nm->promotedFieldsEnabled = ((u64)dsrm->networkMessageContentMask &
                                  (u64)UA_UADPNETWORKMESSAGECONTENTMASK_PROMOTEDFIELDS) != 0;
+    /* Set the SecurityHeader */
+#ifdef UA_ENABLE_PUBSUB_ENCRYPTION
+    if(readerGroup->config.securityMode > UA_MESSAGESECURITYMODE_NONE) {
+        nm->securityEnabled = true;
+        nm->securityHeader.networkMessageSigned = true;
+        if(readerGroup->config.securityMode >= UA_MESSAGESECURITYMODE_SIGNANDENCRYPT)
+            nm->securityHeader.networkMessageEncrypted = true;
+        nm->securityHeader.securityTokenId = readerGroup->securityTokenId;
+
+        /* Generate the MessageNonce */
+        UA_ByteString_allocBuffer(&nm->securityHeader.messageNonce, 8);
```

Comment:
An additional malloc for every generated network message looks like a lot of overhead.
Avoid malloc in the RT path!

You know this is 8byte long. So could be statically allocated and/or reused between network messages.